File: KITE-Dashboard/src/org/webrtc/kite/pojo/Score.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.webrtc.kite.pojo;

import org.webrtc.kite.OverviewResult;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;

/**
 * Score object containing the information as test name, map of browser list and simplified results.
 */
public class Score {

  private String testName;
  private String description;
  private HashMap<Browser,String> score;
  private int tupleSize;

  /**
   * Constructs a new Score from given information as test name, browser list and simplified results.
   *
   * @param test the test.
   * @param overviewResult list of result to construct the score map.
   */
  public Score(ConfigTest test, OverviewResult overviewResult) {
    this.testName = test.getTestName();
    this.description = test.getDescription();
    this.score = processResult(overviewResult.getListOfResultTable());
    this.tupleSize = overviewResult.getTupleSize();
  }


  /**
   * Returns a non repetitive map of all browsers with corresponded simplified result
   * which are SUCCESSFUL, FAILED, or ELSE.
   *
   * @param listOfResult list of results.
   */
  private HashMap<Browser,String> processResult(List<ResultTable> listOfResult){
    HashMap<Browser,String> res = new LinkedHashMap<>();
    List<Browser> browsers = getDistincBrowserList(listOfResult);
    List<ResultTable> tmp = new ArrayList<>(listOfResult);
    for (Browser browser: browsers){
      int total=0, success=0;
      String resTmp="";
      boolean hasSUCCESSFUL = false;
      boolean hasFAILED = false;
      for (ResultTable result: tmp){
        if (result.hasBrowser(browser)){
          total+=1;
          if (result.getResult().equalsIgnoreCase("SUCCESSFUL")) {
            hasSUCCESSFUL = true;
            success+=1;
          } else
            hasFAILED=true;
          //listOfResult.remove(result);
        }
      }
      if (hasFAILED&&hasSUCCESSFUL)
        resTmp="ELSE";
      if (hasFAILED&&!hasSUCCESSFUL)
        resTmp="FAILED";
      if (!hasFAILED&&hasSUCCESSFUL)
        resTmp="SUCCESSFUL";
      resTmp+="_"+success+"/"+total;
      res.put(browser,resTmp);
      tmp = new ArrayList<>(listOfResult);
    }
    return res;
  }

  /**
   * Returns a non repetitive list of all browsers in a result list.
   *
   * @param listOfResult list of results.
   */
  private List<Browser> getDistincBrowserList(List<ResultTable> listOfResult){
    List<Browser> res = new ArrayList<>();
    for (ResultTable result: listOfResult){
      for (Browser browser: result.getBrowserList())
        if (!res.contains(browser))
          res.add(browser);
    }
    return res;
  }

  /**
   * Returns under JSON format the scoreboard
   *
   */
  public String getScoreJson(){
    String res = "";
    res+="[";
    List<Browser> browsers= new ArrayList<>(score.keySet());
    for (int i=0;i<browsers.size(); i++){
      Browser browser = browsers.get(i);
      res+="{";
      res+="\"description\":"+"\""+this.description+"\",";
      res+="\"name\":"+"\""+browser.getName()+"\",";
      res+="\"version\":"+"\""+browser.getVersion()+"\",";
      res+="\"platform\":"+"\""+browser.getPlatform()+"\",";
      String tmp = score.get(browser);
      String result = tmp.split("_")[0];
      String ratio = tmp.split("_")[1];
      res+="\"result\":"+"\""+result+"\",";
      res+="\"ratio\":"+"\""+ratio+"\",";
      res+="\"percentage\":"+(100*Integer.parseInt(ratio.split("/")[0])/Integer.parseInt(ratio.split("/")[1]));
      res+="}";
      if (i<browsers.size()-1)
        res+=",";
    }
    res+="]";
    return res;
  }

  public HashMap<Browser, String> getScore() {
    return score;
  }

  public int getTupleSize() {
    return tupleSize;
  }

  public String getTestName() {
    return testName;
  }

  public String getDescription() {
    return description;
  }
}

File: KITE-Dashboard/WebContent/assets/js/score.js
//
//	Copyright 2017 Google Inc.
//
//	Licensed under the Apache License, Version 2.0 (the "License");
//	you may not use this file except in compliance with the License.
//	You may obtain a copy of the License at
//
//	    https://www.apache.org/licenses/LICENSE-2.0
//
//	Unless required by applicable law or agreed to in writing, software
//	distributed under the License is distributed on an "AS IS" BASIS,
//	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//	See the License for the specific language governing permissions and
//	limitations under the License.
//

/**
* Display the scoreboard unto container;
*/
function displayScoreboard(){

    var scoreHTML='';
    var scoreMap = new Map();
    var ratioMap = new Map();
    var percentageMap = new Map();
    var descriptionMap = new Map();
    var tests = Object.keys(scoreJson);
    var longestBrowserScoreArr;
    var size = 0;
    tests.forEach(function(test){
        var browserScoreArr = scoreJson[test];
        browserScoreArr.forEach(function(browserScore){
        var key = test + browserScore.name + browserScore.version + browserScore.platform;
            scoreMap.set(key, browserScore.result);
            ratioMap.set(key, browserScore.ratio);
            percentageMap.set(key, browserScore.percentage);
            descriptionMap.set(test, browserScore.description)
        });
        if (browserScoreArr.length>size){
            longestBrowserScoreArr = browserScoreArr;
        }
    });
    scoreHTML+=getBrowserHTML(longestBrowserScoreArr);
    tests.forEach(function(testName){
        scoreHTML+=getTestScore(percentageMap,ratioMap,scoreMap,longestBrowserScoreArr,testName,descriptionMap.get(testName));
    });
    $("#scoreboard").html(scoreHTML);
}

function getTestScore(percentageMap,ratioMap,scoreMap,browserArr,testName,description){
    var html='<tr><td><a href="public" class="test-link" data-toggle="tooltip" data-placement="top" title="'+description+'">'+testName+'</a></td>';
    //var html='<tr><td><a href="public?testname='+testName+'" class="status">'+testName+'</a></td>'
    browserArr.forEach(function(browser){
        var key = testName + browser.name + browser.version + browser.platform;
        var result = scoreMap.get(key);
        var ratio = ratioMap.get(key);
        var percentage = percentageMap.get(key);
        html += '<td align="center" class="status" style="color:white;background:';
        if (result){
            switch(result){
                case 'SUCCESSFUL':
                    html+='green"><i class="pe-7s-check"></i>';
                    break;
                case 'FAILED':
                    html+='red"><i class="pe-7s-close-circle"></i>';
                    break;
                case 'ELSE':
                    html+='gold; opacity:'+percentage/100+'"><i class="pe-7s-help1"></i>';
                    break;
            }
            html+= ratio+'</td>';
        }
        else {
            html+='<i class="pe-7s-power grey"></i></td>';
        }
    });
    html += '</tr>'
    return html;
}

function getBrowserHTML(browserArr){
    var html='<tr><td></td>'
    browserArr.forEach(function(browser){
        html += '<td align="center">';
        switch(browser.name){
            case 'safari':
                html+='<h4 class="status" data-toggle="tooltip" data-placement="top" title=" Version: '+browser.version+'"><img src=\"assets/img/safari.png\" height=\"25\" width=\"25\">';
                break;
            case 'firefox':
                html+='<h4 class="status" data-toggle="tooltip" data-placement="top" title=" Version: '+browser.version+'"><img src=\"assets/img/firefox.png\" height=\"25\" width=\"25\">';
                break;
            case 'chrome':
                html+='<h4 class="status" data-toggle="tooltip" data-placement="top" title=" Version: '+browser.version+'"><img src=\"assets/img/chrome.png\" height=\"25\" width=\"25\">';
                break;
            case 'MicrosoftEdge':
                html+='<h4 class="status" data-toggle="tooltip" data-placement="top" title=" Version: '+browser.version+'"><img src=\"assets/img/edge.png\" height=\"25\" width=\"25\">';
                break;
        }
        html+='</h4>';
        html+='<h4 class="status">'+browser.platform+'</h4>';
        html+='</td>';
    });
    html += '</tr>'
    return html;
}

$(document).ready(function(){
    $('[data-toggle="tooltip"]').tooltip();
});